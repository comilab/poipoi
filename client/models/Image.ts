import dayjs, { Dayjs } from 'dayjs'

export default class Image {
  id: number
  postId: number
  filename: string
  extension: string
  weight: number
  createdAt: Dayjs|null = null
  updatedAt: Dayjs|null = null
  publicPaths: {
    original: string
    large: string
    medium: string
    small: string
  }

  constructor (data: any) {
    this.id = data.id
    this.postId = data.postId
    this.filename = data.filename
    this.extension = data.extension
    this.weight = data.weight
    this.publicPaths = data.publicPaths
    if (data.createdAt) {
      this.createdAt = dayjs.utc(data.createdAt).local()
    }
    if (data.updatedAt) {
      this.updatedAt = dayjs.utc(data.updatedAt).local()
    }
  }

  get paths () {
    return {
      original: `${process.env.apiUrl}${this.publicPaths.original}`,
      large: `${process.env.apiUrl}${this.publicPaths.large}`,
      medium: `${process.env.apiUrl}${this.publicPaths.medium}`,
      small: `${process.env.apiUrl}${this.publicPaths.small}`
    }
  }
}
