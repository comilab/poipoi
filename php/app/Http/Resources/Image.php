<?php

namespace App\Http\Resources;

use Illuminate\Http\Resources\Json\JsonResource;

class Image extends JsonResource
{
    /**
     * Transform the resource into an array.
     *
     * @param  \Illuminate\Http\Request  $request
     * @return array
     */
    public function toArray($request)
    {
        return [
            'id' => $this->id,
            'userId' => $this->user_id,
            'postId' => $this->post_id,
            'filename' => $this->filename,
            'extension' => $this->extension,
            'weight' => $this->weight,
            'publicPaths' => $this->public_paths,
            'createdAt' => $this->created_at,
            'updatedAt' => $this->updated_at,
        ];
    }
}
