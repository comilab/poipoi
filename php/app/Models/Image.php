<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Support\Facades\Storage;

class Image extends Model
{
    use HasFactory;

    protected $casts = [
        'post_id' => 'integer',
        'weight' => 'integer',
    ];

    protected $touches = ['post'];

    public function post()
    {
        return $this->belongsTo(Post::class);
    }

    public function getPathsAttribute(): array
    {
        return [
            'original' => "public/images/{$this->post_id}/original_{$this->filename}.{$this->extension}",
            'large' => "public/images/{$this->post_id}/large_{$this->filename}.jpeg",
            'medium' => "public/images/{$this->post_id}/medium_{$this->filename}.jpeg",
            'small' => "public/images/{$this->post_id}/small_{$this->filename}.jpeg",
        ];
    }

    public function getPublicPathsAttribute(): array
    {
        return [
            'original' => Storage::url($this->paths['original']),
            'large' => Storage::url($this->paths['large']),
            'medium' => Storage::url($this->paths['medium']),
            'small' => Storage::url($this->paths['small']),
        ];
    }

    public function saveImage(string $tempFile)
    {
        Storage::move("{$tempFile}", $this->paths['original']);
        Storage::move("{$tempFile}.large", $this->paths['large']);
        Storage::move("{$tempFile}.medium", $this->paths['medium']);
        Storage::move("{$tempFile}.small", $this->paths['small']);
    }

    public function deleteImage()
    {
        Storage::delete($this->paths['original']);
        Storage::delete($this->paths['large']);
        Storage::delete($this->paths['medium']);
        Storage::delete($this->paths['small']);
    }
}
