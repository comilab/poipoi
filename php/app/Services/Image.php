<?php

namespace App\Services;

use Intervention\Image\Facades\Image as FacadesImage;

class Image
{
    public function resize(string $from, string $to, int $width, int $height)
    {
        FacadesImage::make($from)
            ->resize($width, $height, function ($constraint) {
                $constraint->aspectRatio();
                $constraint->upsize();
            })
            ->save($to, 90, 'jpeg');
    }
}
