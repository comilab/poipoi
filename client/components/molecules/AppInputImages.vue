<template lang="pug">
.text-center
  .bg-gray-200.rounded.my-4.p-4
    ValidationProvider(
      :rules="{ max: 200 }",
      v-slot="{ invalid }",
      slim
    )
      Draggable.grid.gap-8.grid-cols-2.md_grid-cols-4.mb-1(
        v-if="innerImages.length",
        v-model="innerImages"
      )
        AppInputImagesThumbnail(
          v-for="(image, i) in innerImages",
          :key="i",
          :image="image",
          :num="i + 1",
          @remove="onRemove(image)"
        )
      .text-sm.text-right(:class="{ 'text-red-700': invalid }") {{ innerImages.length }} / 200
  .grid.md_grid-cols-2.md_divide-x.items-stretch.my-4
    .activator.flex.justify-center
      VueUploadComponent(
        ref="uploader",
        v-model="uploadImages",
        :multiple="true",
        :extensions="/\.(gif|jpe?g|png)$/i",
        accept="image/*",
        :drop="true",
        :drop-directory="false",
        :post-action="`${postAction}/images`",
        :headers="headers",
        @input-filter="onInputFilter"
      )
        AppButton(
          color="blue",
          :disabled="innerImages.length >= 200"
        )
          div ここをクリックして画像を選択
          .hidden.lg_block
            .text-sm もしくは
            div 画像をドラッグ&amp;ドロップ
    .flex.justify-center.mt-4.md_mt-0
      .inline-flex.items-center.bg-blue-600.text-white.hover_bg-blue-800.rounded.cursor-pointer.h-full.px-4.py-2(
        :contenteditable="!innerImages.length >= 200",
        :class="{ 'opacity-50': innerImages.length >= 200 }",
        @paste.prevent="onPaste",
        role="textbox",
        aria-label="ここをクリックして画像をペースト"
      ) ここをクリックして画像をペースト
</template>

<script lang="ts">
import { defineComponent, ref, computed } from 'nuxt-composition-api'
import { PropType, watch } from '@vue/composition-api'
import VueUploadComponent from 'vue-upload-component'
// @ts-ignore
import Draggable from 'vuedraggable'

import Image from '~/models/Image'
import AppButton from '~/components/atoms/AppButton.vue'
import AppInputImagesThumbnail from '~/components/molecules/AppInputImagesThumbnail.vue'

interface WeightedVUFile extends VUFile {
  weight: number
}

export default defineComponent({
  components: {
    VueUploadComponent,
    Draggable,
    AppButton,
    AppInputImagesThumbnail
  },
  props: {
    images: {
      type: Array as PropType<Image[]>,
      required: false,
      default: () => []
    }
  },
  setup (props, context) {
    const uploader = ref<VueUploadComponent>()

    const innerImages = ref<(WeightedVUFile|Image)[]>([...props.images])

    const uploadImages = computed({
      get: () => {
        return innerImages.value
          .map((file, i) => {
            file.weight = i
            return file
          })
          .filter(file => !(file instanceof Image)) as WeightedVUFile[]
      },
      set: (newFiles) => {
        innerImages.value = innerImages.value
          .map((file, i) => {
            if (file instanceof Image) {
              return file
            }
            return newFiles.find(file => file.weight === i)
          })
          .filter(file => !!file)
          .concat(newFiles.filter(file => file.weight === undefined)) as (WeightedVUFile|Image)[]
      }
    })

    const headers = computed(() => {
      const value = document.cookie.split('; ').find((segment) => {
        return segment.startsWith('XSRF-TOKEN')
      })!.split('=').pop()

      return {
        'X-XSRF-TOKEN': decodeURIComponent(value!)
      }
    })

    const onRemove = (target: WeightedVUFile|Image) => {
      if (target instanceof Image) {
        const key = innerImages.value.findIndex(file => file === target)
        innerImages.value.splice(key, 1)
      } else {
        uploader.value!.remove(target)
      }
    }

    const onPaste = (event: ClipboardEvent) => {
      if (event.clipboardData?.items.length) {
        const item = event.clipboardData.items[0]
        if (item && item.kind === 'file' && /^image\//.test(item.type)) {
          uploader.value!.add(item.getAsFile()!)
        }
      }
    }

    const onInputFilter = (newFile: VUFile) => {
      if (newFile && newFile.xhr && !newFile.xhr.withCredentials) {
        newFile.xhr.withCredentials = true
      }
    }

    const upload = () => {
      if (uploadImages.value.length) {
        uploader.value!.active = true
      } else {
        context.emit('uploaded', innerImages)
      }
    }

    watch(
      () => !!uploader.value?.uploaded,
      (isUploaded) => {
        if (uploadImages.value.length && isUploaded) {
          context.emit('uploaded', innerImages.value.map((file) => {
            if (!(file instanceof Image)) {
              return file.response
            }
            return file
          }))
        }
      }
    )

    return {
      uploader,
      innerImages,
      uploadImages,
      headers,
      onRemove,
      onPaste,
      onInputFilter,
      upload,
      postAction: `${process.env.apiUrl}/api`
    }
  }
})
</script>

<style lang="postcss" scoped>
.activator >>> label {
  @apply cursor-pointer;
}
</style>
