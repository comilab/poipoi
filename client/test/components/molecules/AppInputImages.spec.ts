import { fireEvent, render, waitFor } from '@testing-library/vue'
import { defineComponent, ref } from 'nuxt-composition-api'
import { Ref } from '@vue/composition-api'

import AppInputImages from '~/components/molecules/AppInputImages.vue'
import Image from '~/models/Image'
import imageFactory from '~/test/factories/image'
import useMockedComponent from '~/test/stubs/MockedComponent'

describe('components/molecules/AppInputImages', () => {
  let options: any
  let props: any

  beforeEach(() => {
    props = {
      value: []
    }

    options = {
      props,
      stubs: {
        FontAwesomeIcon: true
      }
    }

    jest.spyOn(document, 'cookie', 'get').mockImplementation(() => 'XSRF-TOKEN=foo')
  })

  test('renders correctly', () => {
    const { container } = render(AppInputImages, options)

    expect(container).toMatchSnapshot()
  })

  describe('images', () => {
    test('props.imagesを初期値としてセット', () => {
      props.images = imageFactory.buildList(1)
      const { getAllByRole } = render(AppInputImages, options)

      expect(getAllByRole('img')).toHaveLength(1)
    })

    test('Draggableに渡す', () => {
      const { component, propsList } = useMockedComponent({ props: ['value'] })
      options.stubs.Draggable = component
      props.images = imageFactory.buildList(1)
      render(AppInputImages, options)

      expect(propsList[0].value).toStrictEqual(props.images)
    })

    test('空配列の場合はサムネイル一覧を表示しない', () => {
      props.images = []
      const { queryByRole } = render(AppInputImages, options)

      expect(queryByRole('img')).not.toBeInTheDocument()
    })

    test('カウンターを表示', () => {
      props.images = imageFactory.buildList(1)
      const { getByText } = render(AppInputImages, options)

      expect(getByText('1 / 200')).toBeInTheDocument()
    })

    test('画像が200枚以上ならボタン類を使用不可にする', () => {
      props.images = imageFactory.buildList(200)
      options.stubs.AppInputImagesThumbnail = true
      const { getByRole, getByLabelText } = render(AppInputImages, options)

      expect(getByRole('button', { name: /ここをクリックして画像を選択/ })).toHaveAttribute('disabled')
      expect(getByLabelText('ここをクリックして画像をペースト')).toHaveAttribute('contenteditable', 'false')
    })
  })

  describe('AppInputImagesThumbnail', () => {
    let inputImagesThumbnailProps: any
    let inputImagesThumbnailFireRemoveFunction: Function
    let draggableProps: any

    beforeEach(() => {
      options.stubs.AppInputImagesThumbnail = defineComponent({
        props: ['image', 'num'],
        setup (props, { emit }) {
          inputImagesThumbnailProps = props
          inputImagesThumbnailFireRemoveFunction = () => emit('remove')
        },
        template: '<div />'
      })
      options.stubs.Draggable = defineComponent({
        props: ['value'],
        setup (props) {
          draggableProps = props
        },
        template: '<div><slot /></div>'
      })
    })

    test('imagesを1件ずつ渡す', () => {
      props.images = imageFactory.buildList(1)
      render(AppInputImages, options)

      expect(inputImagesThumbnailProps.image).toBe(props.images[0])
    })

    test('num = key + 1', () => {
      props.images = imageFactory.buildList(1)
      render(AppInputImages, options)

      expect(inputImagesThumbnailProps.num).toBe(1)
    })

    describe('remove', () => {
      test('対象がImageモデルの場合は削除', async () => {
        props.images = imageFactory.buildList(1)
        render(AppInputImages, options)
        inputImagesThumbnailFireRemoveFunction()

        await waitFor(() => {
          expect(draggableProps.value).toHaveLength(0)
        })
      })

      test('対象がImageモデルでない場合はuploader.removeを実行', () => {
        const removeMock = jest.fn()
        options.stubs.VueUploadComponent = defineComponent({
          setup () {
            return {
              remove: removeMock
            }
          },
          template: '<div />'
        })
        props.images = [{}]
        render(AppInputImages, options)
        inputImagesThumbnailFireRemoveFunction()

        expect(removeMock).toBeCalled()
      })
    })
  })

  describe('VueUploadComponent', () => {
    let uploadComponentProps: any
    let uploadComponentFireInputFunction: Function
    let uploadComponentFireInputFilterFunction: Function
    let draggableProps: any

    beforeEach(() => {
      options.stubs.VueUploadComponent = defineComponent({
        props: ['value', 'multiple', 'extensions', 'accept', 'drop', 'dropDirectory', 'postAction', 'headers'],
        setup (props, { emit }) {
          uploadComponentProps = props
          uploadComponentFireInputFunction = (arg: any[]) => emit('input', arg)
          uploadComponentFireInputFilterFunction = (arg: any) => emit('input-filter', arg)
        },
        template: '<div />'
      })
      options.stubs.Draggable = defineComponent({
        props: ['value'],
        setup (props) {
          draggableProps = props
        },
        template: '<div />'
      })
    })

    test('valueはimagesからImageモデルを除外したもの', () => {
      props.images = [imageFactory.build(), {}]
      render(AppInputImages, options)

      expect(uploadComponentProps.value).toHaveLength(1)
      expect(uploadComponentProps.value[0]).not.toBeInstanceOf(Image)
    })

    test('multiple = true', () => {
      render(AppInputImages, options)

      expect(uploadComponentProps.multiple).toBeTruthy()
    })

    test('extensions = /.(gif|jpe?g|png)$/i', () => {
      render(AppInputImages, options)

      expect(uploadComponentProps.extensions).toBeInstanceOf(RegExp)
      expect(uploadComponentProps.extensions.test('.gif')).toBeTruthy()
      expect(uploadComponentProps.extensions.test('.jpeg')).toBeTruthy()
      expect(uploadComponentProps.extensions.test('.jpg')).toBeTruthy()
      expect(uploadComponentProps.extensions.test('.png')).toBeTruthy()
    })

    test('accept = image/*', () => {
      render(AppInputImages, options)

      expect(uploadComponentProps.accept).toBe('image/*')
    })

    test('drop = true', () => {
      render(AppInputImages, options)

      expect(uploadComponentProps.drop).toBeTruthy()
    })

    test('dropDirectory = false', () => {
      render(AppInputImages, options)

      expect(uploadComponentProps.dropDirectory).toBeFalsy()
    })

    test('postAction = /api/images', () => {
      render(AppInputImages, options)

      expect(uploadComponentProps.postAction).toBe(`${process.env.apiUrl}/api/images`)
    })

    test('headersにX-XSRF-TOKENをセット', () => {
      render(AppInputImages, options)

      expect(uploadComponentProps.headers).toStrictEqual({ 'X-XSRF-TOKEN': 'foo' })
    })

    describe('input', () => {
      test('imagesを更新', async () => {
        props.images = []
        render(AppInputImages, options)
        uploadComponentFireInputFunction([{}])

        await waitFor(() => {
          expect(draggableProps.value).toHaveLength(1)
        })
      })

      test('ファイルが削除された場合に対応', async () => {
        props.images = [{ id: 1, weight: 0 }, { id: 2, weight: 1 }]
        render(AppInputImages, options)
        uploadComponentFireInputFunction([{ id: 1, weight: 0 }])

        await waitFor(() => {
          expect(draggableProps.value).toHaveLength(1)
          expect(draggableProps.value[0].id).toBe(1)
        })
      })

      test('Imageモデルはそのまま保持', async () => {
        const image = imageFactory.build()
        props.images = [image]
        render(AppInputImages, options)
        uploadComponentFireInputFunction([image, {}])

        await waitFor(() => {
          expect(draggableProps.value).toHaveLength(2)
          expect(draggableProps.value[0]).toStrictEqual(image)
        })
      })
    })

    describe('input-filter', () => {
      test('newFile.xhr.withCredentialsをセット', () => {
        const newFile = {
          xhr: {
            withCredentials: false
          }
        }
        render(AppInputImages, options)
        uploadComponentFireInputFilterFunction(newFile)

        expect(newFile.xhr.withCredentials).toBeTruthy()
      })
    })
  })

  describe('onPaste', () => {
    let event: any
    let getAsFileMock: jest.Mock
    let uploaderComponentAddMock: jest.Mock

    beforeEach(() => {
      getAsFileMock = jest.fn().mockReturnValue('file')
      event = {
        clipboardData: {
          items: [
            { kind: 'file', type: 'image/jpeg', getAsFile: getAsFileMock }
          ]
        }
      }
      uploaderComponentAddMock = jest.fn()
      options.stubs.VueUploadComponent = defineComponent({
        setup () {
          return {
            add: uploaderComponentAddMock
          }
        },
        template: '<div />'
      })
    })

    test('uploader.addを実行', async () => {
      const { getByLabelText } = render(AppInputImages, options)
      await fireEvent.paste(getByLabelText('ここをクリックして画像をペースト'), event)

      expect(getAsFileMock).toBeCalled()
      expect(uploaderComponentAddMock).toBeCalledWith('file')
    })

    test('画像以外がペーストされた場合は何もしない', async () => {
      event.clipboardData.items[0].type = 'text/html'
      const { getByLabelText } = render(AppInputImages, options)
      await fireEvent.paste(getByLabelText('ここをクリックして画像をペースト'), event)

      expect(getAsFileMock).not.toBeCalled()
      expect(uploaderComponentAddMock).not.toBeCalled()
    })
  })

  describe('upload', () => {
    let wrapper: any
    let images: any[]
    let onUploaded: jest.Mock
    let appInputImages: Ref<any>

    beforeEach(() => {
      onUploaded = jest.fn()
      wrapper = defineComponent({
        components: { AppInputImages },
        setup () {
          appInputImages = ref()
          return {
            appInputImages,
            images,
            onUploaded
          }
        },
        template: `
          <div>
            <AppInputImages
              ref="appInputImages"
              :images="images"
              @uploaded="onUploaded"
            />
          </div>
        `
      })
      options.stubs.VueUploadComponent = defineComponent({
        setup () {
          return {
            active: false
          }
        },
        template: '<div />'
      })
    })

    test('uploadImagesがあればuploader.active=trueにする', () => {
      images = [{}]
      render(wrapper, options)
      appInputImages.value.upload()

      expect(appInputImages.value.uploader.active).toBeTruthy()
    })

    test('uploadImagesがない場合はuploadedを発火', () => {
      images = imageFactory.buildList(1)
      render(wrapper, options)
      appInputImages.value.upload()

      expect(onUploaded).toBeCalledWith({ value: images })
    })
  })

  describe('watch', () => {
    describe('uploaded', () => {
      let uploadComponentSetUploadedFunction: Function

      beforeEach(() => {
        options.stubs.VueUploadComponent = defineComponent({
          setup (_props, { root }) {
            const uploaded = ref(false)
            uploadComponentSetUploadedFunction = async () => {
              uploaded.value = true
              await root.$nextTick()
            }
            return {
              uploaded
            }
          },
          template: '<div />'
        })
      })

      test('trueになったらuploadedイベントを発火', async () => {
        props.images = [{
          response: {}
        }]
        const { emitted } = render(AppInputImages, options)
        await uploadComponentSetUploadedFunction()

        expect(emitted().uploaded).toHaveLength(1)
        expect(emitted().uploaded[0]).toHaveLength(1)
        expect(emitted().uploaded[0][0]).toHaveLength(1)
      })

      test('Imageモデルはそのまま渡す', async () => {
        props.images = [
          imageFactory.build(),
          {
            response: {}
          }
        ]
        const { emitted } = render(AppInputImages, options)
        await uploadComponentSetUploadedFunction()

        expect(emitted().uploaded[0][0][0]).toBe(props.images[0])
      })

      test('Imageモデルでない場合はresponseを渡す', async () => {
        props.images = [{
          response: {}
        }]
        const { emitted } = render(AppInputImages, options)
        await uploadComponentSetUploadedFunction()

        expect(emitted().uploaded[0][0][0]).toBe(props.images[0].response)
      })

      test('uploadImagesがなければ何もしない', async () => {
        props.images = []
        const { emitted } = render(AppInputImages, options)
        await uploadComponentSetUploadedFunction()

        expect(emitted().uploaded).toBeFalsy()
      })
    })
  })
})
