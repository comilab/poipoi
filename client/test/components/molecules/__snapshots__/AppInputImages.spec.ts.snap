// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/molecules/AppInputImages renders correctly 1`] = `
<div>
  <div
    class="text-center"
    value=""
  >
    <div
      class="bg-gray-200 rounded my-4 p-4"
    >
      <span>
        <!---->
        <div
          class="text-sm text-right"
        >
          0 / 200
        </div>
      </span>
    </div>
    <div
      class="grid md_grid-cols-2 md_divide-x items-stretch my-4"
    >
      <div
        class="activator flex justify-center"
      >
        <span
          class="file-uploads file-uploads-html5"
        >
          <button
            class="inline-flex items-center justify-center whitespace-no-wrap transition-colors duration-300 disabled_opacity-75 disabled_pointer-events-none focus_outline-none bg-blue-600 text-white hover_bg-blue-700 px-3 py-2 rounded"
            data-testid="button"
            type="button"
          >
            <!---->
            <div
              class=""
              data-testid="text"
            >
              <div>
                ここをクリックして画像を選択
              </div>
              <div
                class="hidden lg_block"
              >
                <div
                  class="text-sm"
                >
                  もしくは
                </div>
                <div>
                  画像をドラッグ&ドロップ
                </div>
              </div>
            </div>
          </button>
           
          <label
            for="file"
          />
           
          <input
            accept="image/*"
            id="file"
            multiple="multiple"
            name="file"
            type="file"
          />
        </span>
      </div>
      <div
        class="flex justify-center mt-4 md_mt-0"
      >
        <div
          aria-label="ここをクリックして画像をペースト"
          class="inline-flex items-center bg-blue-600 text-white hover_bg-blue-800 rounded cursor-pointer h-full px-4 py-2"
          contenteditable="false"
          role="textbox"
        >
          ここをクリックして画像をペースト
        </div>
      </div>
    </div>
  </div>
</div>
`;
